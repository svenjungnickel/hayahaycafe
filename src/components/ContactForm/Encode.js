const Encode = (data) => {
    if ('object' !== typeof data) {
        return;
    }

    return Object.keys(data)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(data[key]))
        .join('&');
};

export default Encode;
