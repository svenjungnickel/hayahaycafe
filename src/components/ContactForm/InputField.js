import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-bootstrap';
import ContactFormStyles from '../../styles/components/ContactForm/ContactForm.module.scss';

const InputField = ({ controlId, name, label, errorMessage, onChange, type, description, ariaLabel }) => (
    <Form.Group controlId={controlId} data-cy={controlId}>
        {'' !== label && <Form.Label>{label} *</Form.Label>}
        <Form.Control name={name} type={type} required onChange={onChange} aria-label={ariaLabel} />
        {description && (
            <Form.Text className={`text-muted ${ContactFormStyles.contactFormTextMuted}`}>{description}</Form.Text>
        )}
        <Form.Control.Feedback type="invalid">{errorMessage}</Form.Control.Feedback>
    </Form.Group>
);

InputField.defaultProps = {
    type: 'text',
};

InputField.propTypes = {
    controlId: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    errorMessage: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    type: PropTypes.oneOf(['email', 'password', 'text']),
    description: PropTypes.string,
    ariaLabel: PropTypes.string,
};

export default InputField;
