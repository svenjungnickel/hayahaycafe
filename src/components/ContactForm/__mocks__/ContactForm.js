import React from 'react';

const ContactForm = () => <form />;

export default ContactForm;
