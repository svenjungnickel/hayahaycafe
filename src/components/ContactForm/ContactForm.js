import React, { useState, createRef } from 'react';
import { Alert, Button, Col, Form } from 'react-bootstrap';
import Recaptcha from 'react-google-recaptcha';
import Fade from 'react-reveal/Fade';
import InputField from './InputField';
import ContactFormStyles from '../../styles/components/ContactForm.module.scss';

const RECAPTCHA_KEY = process.env.SITE_RECAPTCHA_KEY;
const CONTACT_FORM_FADE_UP_DURATION = 400;
const SUCCESS_MESSAGE_FADE_IN_DURATION = 200;

const encode = data => {
    return Object.keys(data)
        .map(key => encodeURIComponent(key) + '=' + encodeURIComponent(data[key]))
        .join('&');
};

const ContactForm = () => {
    const [validated, setValidated] = useState(false);
    const [state, setState] = useState({});
    const [recaptchaError, setRecaptchaError] = useState(false);
    const [success, setSuccess] = useState(false);
    const recaptchaRef = createRef();

    const handleChange = e => {
        setState({ ...state, [e.target.name]: e.target.value });
    };

    const handleRecaptcha = () => {
        setRecaptchaError(false);
    };

    const handleSubmit = event => {
        event.preventDefault();
        event.stopPropagation();
        setValidated(true);

        const recaptchaValue = recaptchaRef.current.getValue() || false;
        if (false === recaptchaValue) {
            setRecaptchaError(true);
            return;
        }

        if (event.currentTarget.checkValidity() === false) {
            return;
        }

        const form = event.target;
        fetch('/', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: encode({
                'form-name': form.getAttribute('name'),
                'g-recaptcha-response': recaptchaValue,
                ...state,
            }),
        })
            .then(() => {
                setSuccess(true);
            })
            .catch(error => alert(error));
    };

    return (
        <div className={ContactFormStyles.contactForm}>
            <Fade top collapse when={false === success} duration={CONTACT_FORM_FADE_UP_DURATION} unmountOnExit={true}>
                <Form
                    noValidate
                    validated={validated}
                    onSubmit={handleSubmit}
                    name="contact"
                    method="POST"
                    action="/thanks"
                    data-netlify="true"
                    data-netlify-recaptcha="true"
                >
                    <Form.Label htmlFor="contactFirstName">Name *</Form.Label>
                    <Form.Row>
                        <Col xs={12} lg={6}>
                            <InputField
                                controlId="contactFirstName"
                                name="firstName"
                                label=""
                                errorMessage="Please enter your first name."
                                onChange={handleChange}
                                description="First name"
                                ariaLabel="First name"
                            />
                        </Col>
                        <Col xs={12} lg={6}>
                            <InputField
                                controlId="contactLastName"
                                name="lastName"
                                label=""
                                errorMessage="Please enter your last name."
                                onChange={handleChange}
                                description="Last name"
                                ariaLabel="Last name"
                            />
                        </Col>
                    </Form.Row>

                    <InputField
                        controlId="contactEmail"
                        name="email"
                        label="Email"
                        errorMessage="Please enter a valid email address."
                        onChange={handleChange}
                    />

                    <InputField
                        controlId="contactSubject"
                        name="subject"
                        label="Subject"
                        errorMessage="Please enter a subject."
                        onChange={handleChange}
                    />

                    <Form.Group controlId="contactMessage">
                        <Form.Label>Message *</Form.Label>
                        <textarea
                            name="message"
                            className="form-control"
                            id="contactMessage"
                            rows="5"
                            required
                            onChange={handleChange}
                        />
                        <Form.Control.Feedback type="invalid">Please enter your message.</Form.Control.Feedback>
                    </Form.Group>

                    <Form.Group controlId="contactRecaptcha">
                        <Recaptcha
                            id="contactRecaptcha"
                            ref={recaptchaRef}
                            sitekey={RECAPTCHA_KEY}
                            onChange={handleRecaptcha}
                            className={ContactFormStyles.contactFormReCaptcha}
                        />
                        {true === recaptchaError && (
                            <div className="invalid-feedback" style={{ display: 'block' }}>
                                Please answer the captcha.
                            </div>
                        )}
                    </Form.Group>

                    <Button variant="primary" type="submit" className={ContactFormStyles.contactFormSubmitButton}>
                        Submit
                    </Button>
                </Form>
            </Fade>
            <Fade
                collapse
                when={success}
                duration={SUCCESS_MESSAGE_FADE_IN_DURATION}
                delay={CONTACT_FORM_FADE_UP_DURATION}
            >
                <Alert variant="success" className={ContactFormStyles.contactFormSuccess}>
                    <p>{"Thank you for your message. We're getting in touch with you soon."}</p>
                </Alert>
            </Fade>
        </div>
    );
};

export default ContactForm;
