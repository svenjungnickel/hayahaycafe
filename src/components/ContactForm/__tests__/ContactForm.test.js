import React, { useState as useStateMock } from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import ContactForm from '../ContactForm';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn(),
}));

describe('ContactForm snapshot', () => {
    beforeEach(() => {
        useStateMock.mockImplementation(init => [init, jest.fn()]);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const component = <ContactForm />;
        const tree = renderer.create(component).toJSON();

        expect(tree).toMatchSnapshot();
    });
});

describe('ContactForm events', () => {
    let ContactFormComponent;
    const setStateMock = jest.fn();
    const setValidatedMock = jest.fn();
    const setRecaptchaErrorMock = jest.fn();
    const setSuccessMock = jest.fn();

    beforeEach(() => {
        useStateMock
            .mockImplementationOnce(init => [init, setValidatedMock])
            .mockImplementationOnce(init => [init, setStateMock])
            .mockImplementationOnce(init => [init, setRecaptchaErrorMock])
            .mockImplementationOnce(init => [init, setSuccessMock]);
        ContactFormComponent = mount(<ContactForm />);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('simulate change input field', () => {
        const InputField = ContactFormComponent.find('#contactFirstName');
        InputField.simulate('change');

        expect(setValidatedMock).toHaveBeenCalledTimes(0);
        expect(setStateMock).toHaveBeenCalledTimes(1);
        expect(setStateMock).toHaveBeenCalledWith({ firstName: '' });
        expect(setRecaptchaErrorMock).toHaveBeenCalledTimes(0);
        expect(setSuccessMock).toHaveBeenCalledTimes(0);
    });

    it('simulate change text area', () => {
        const TextArea = ContactFormComponent.find('#contactMessage');
        TextArea.simulate('change');

        expect(setValidatedMock).toHaveBeenCalledTimes(0);
        expect(setStateMock).toHaveBeenCalledTimes(1);
        expect(setStateMock).toHaveBeenCalledWith({ message: '' });
        expect(setRecaptchaErrorMock).toHaveBeenCalledTimes(0);
        expect(setSuccessMock).toHaveBeenCalledTimes(0);
    });

    // @TODO finish skipped tests
    it.skip('simulate change unsolved recaptcha', () => {
        const Recaptcha = ContactFormComponent.find('#contactRecaptcha').first();
        Recaptcha.simulate('click');

        expect(setValidatedMock).toHaveBeenCalledTimes(0);
        expect(setStateMock).toHaveBeenCalledTimes(0);
        expect(setRecaptchaErrorMock).toHaveBeenCalledTimes(1);
        expect(setRecaptchaErrorMock).toHaveBeenCalledWith(false);
        expect(setSuccessMock).toHaveBeenCalledTimes(0);
    });

    it('simulate unsuccessful form submit', () => {
        const SubmitButton = ContactFormComponent.find('button[type="submit"]');
        SubmitButton.simulate('submit');

        expect(setValidatedMock).toHaveBeenCalledTimes(1);
        expect(setValidatedMock).toHaveBeenCalledWith(true);
        expect(setStateMock).toHaveBeenCalledTimes(0);
        expect(setRecaptchaErrorMock).toHaveBeenCalledTimes(1);
        expect(setSuccessMock).toHaveBeenCalledTimes(0);
    });

    it.skip('simulate successful form submit', () => {
        useStateMock
            .mockImplementationOnce(init => [init, setValidatedMock])
            .mockImplementationOnce(init => [init, setStateMock])
            .mockImplementationOnce(init => [init, setRecaptchaErrorMock])
            .mockImplementationOnce(init => [init, setSuccessMock])
            .mockReturnValue(true);
        // mock setRecaptchaErrorMock return to false
        //...
        ContactFormComponent = mount(<ContactForm />);

        const SubmitButton = ContactFormComponent.find('button[type="submit"]');
        SubmitButton.simulate('submit');
        // validate state mock calls and params
        //...
    });
});
