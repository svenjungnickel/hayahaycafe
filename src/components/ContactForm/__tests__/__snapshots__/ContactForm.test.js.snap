// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm snapshot renders correctly 1`] = `
<div
  className="contactForm"
>
  <div
    style={
      Object {
        "height": 0,
        "visibility": undefined,
      }
    }
  >
    <div
      className="react-reveal"
      style={
        Object {
          "opacity": undefined,
        }
      }
    >
      <form
        action="/thanks"
        className=""
        data-netlify="true"
        data-netlify-recaptcha="true"
        method="POST"
        name="contact"
        noValidate={true}
        onSubmit={[Function]}
      >
        <label
          className="form-label"
          htmlFor="contactFirstName"
        >
          Name *
        </label>
        <div
          className="form-row"
        >
          <div
            className="col-lg-6 col-12"
          >
            <div
              className="form-group"
            >
              <input
                aria-label="First name"
                className="form-control"
                id="contactFirstName"
                name="firstName"
                onChange={[Function]}
                required={true}
                type="text"
              />
              <small
                className="text-muted contactFormTextMuted form-text"
              >
                First name
              </small>
              <div
                className="invalid-feedback"
              >
                Please enter your first name.
              </div>
            </div>
          </div>
          <div
            className="col-lg-6 col-12"
          >
            <div
              className="form-group"
            >
              <input
                aria-label="Last name"
                className="form-control"
                id="contactLastName"
                name="lastName"
                onChange={[Function]}
                required={true}
                type="text"
              />
              <small
                className="text-muted contactFormTextMuted form-text"
              >
                Last name
              </small>
              <div
                className="invalid-feedback"
              >
                Please enter your last name.
              </div>
            </div>
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="contactEmail"
          >
            Email
             *
          </label>
          <input
            className="form-control"
            id="contactEmail"
            name="email"
            onChange={[Function]}
            required={true}
            type="text"
          />
          <div
            className="invalid-feedback"
          >
            Please enter a valid email address.
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="contactSubject"
          >
            Subject
             *
          </label>
          <input
            className="form-control"
            id="contactSubject"
            name="subject"
            onChange={[Function]}
            required={true}
            type="text"
          />
          <div
            className="invalid-feedback"
          >
            Please enter a subject.
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="contactMessage"
          >
            Message *
          </label>
          <textarea
            className="form-control"
            id="contactMessage"
            name="message"
            onChange={[Function]}
            required={true}
            rows="5"
          />
          <div
            className="invalid-feedback"
          >
            Please enter your message.
          </div>
        </div>
        <div
          className="form-group"
        >
          <div
            className="contactFormReCaptcha"
            id="contactRecaptcha"
          />
        </div>
        <button
          className="contactFormSubmitButton btn btn-primary"
          disabled={false}
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
  <div
    style={
      Object {
        "height": 0,
        "visibility": "hidden",
      }
    }
  >
    <div
      className="react-reveal"
      style={
        Object {
          "opacity": 0,
        }
      }
    >
      <div
        className="fade contactFormSuccess alert alert-success show"
        role="alert"
      >
        <p>
          Thank you for your message. We're getting in touch with you soon.
        </p>
      </div>
    </div>
  </div>
</div>
`;
